Begin project

jeanconn opened the pull request:
Begin project.

jeanconn committed:
Import sources

jeanconn committed:
Remove idea for inspecting roll-independent circle

jeanconn committed:
Arrange table columns

jeanconn committed:
Use dashed options for start/stop days

jeanconn committed:
Remove commented-out code

jeanconn committed:
Reduce decimal place printing of data

jeanconn committed:
Don't check day roll ranges if warm temp works

jeanconn committed:
Add a docstring

jeanconn committed:
Fix sort order of table and change output format

jeanconn committed:
Filter agasc in one step and limit columns

jeanconn committed:
Use proper motion corrected positions

jeanconn committed:
Simplify grabbing 8 brightest stars

jeanconn committed:
Work away from nominal roll to find best roll

jeanconn commented on aca_required_temp.py:
```diff
@@ -0,0 +1,199 @@
+                    best_temp = roll_temp
+                    best_roll = roll
+                if best_temp == characteristics.WARM_T_CCD:
+                    break
```
See notes at  https://github.com/sot/aca_required_temp/commit/5aa1c26a5cfb58e42e86771b4e82c139e237f3f4

jeanconn commented on aca_required_temp.py:
```diff
@@ -0,0 +1,199 @@
+    return field
+
+
+def select_fov_stars(ra, dec, roll, field):
```
See https://github.com/sot/aca_required_temp/commit/b4c478d045d22649af08191a9f729c1f8d04e3ee

jeanconn commented on aca_required_temp.py:
```diff
@@ -0,0 +1,199 @@
+                              / 2).date
+
+    # get a list of days
+    day = DateTime(start_day)
```
See https://github.com/sot/aca_required_temp/commit/b4c478d045d22649af08191a9f729c1f8d04e3ee

jeanconn commented on aca_required_temp.py:
```diff
@@ -0,0 +1,199 @@
+
+    temps = {}
+    # loop over them
+    for day in days:
```
See https://github.com/sot/aca_required_temp/commit/b4c478d045d22649af08191a9f729c1f8d04e3ee

jeanconn commented on aca_required_temp.py:
```diff
@@ -0,0 +1,199 @@
+        return None
+    # take the 8 brightest
+    stars = fov_stars[0:8]
+    id_hash = hashlib.md5(stars['AGASC_ID']).hexdigest()
```
See https://github.com/sot/aca_required_temp/commit/1c4fa9b60c05c59ec779094ec08b86b0a3092de2

jeanconn commented on aca_required_temp.py:
```diff
@@ -0,0 +1,199 @@
+                          & (field['CLASS'] == 0)
+                          & (field['ASPQ1'] == 0)
+                          & (field['COLOR1'] != 0.7)
+                          & (field['COLOR1'] != 1.5)][cols]
```
See https://github.com/sot/aca_required_temp/commit/814823cf6f85a55d6e5fdab7ddb73175d2a3ba25

jeanconn commented on aca_required_temp.py:
```diff
@@ -0,0 +1,199 @@
+                        type=float)
+    parser.add_argument("--out",
+                        default="roll_temps.dat")
+    parser.add_argument("--start-day",
```
See https://github.com/sot/aca_required_temp/commit/60d769d82b6a3a667f04900c108082f71caad244

jeanconn commented on aca_required_temp.py:
```diff
@@ -0,0 +1,199 @@
+    rolldev = get_rolldev(day_pitch)
+    # Quat doesn't care about the domain of roll, so I don't have to
+    # do anything tricky with 180/360 etc
+    for rolldiff in np.arange(0, rolldev, step=.5):
```
See https://github.com/sot/aca_required_temp/commit/5aa1c26a5cfb58e42e86771b4e82c139e237f3f4

jeanconn committed:
Bugfix: actually do the filtering in get_agasc_cone

jeanconn committed:
Remove AGASC caching pieces and just pass the stars

jeanconn committed:
Don't filter on color == 1.5

jeanconn committed:
Use a tuple for star combination hash id

jeanconn committed:
Clear star temperature cache appropriately

jeanconn committed:
Convert a bunch of 'temp's to 't_ccd's

jeanconn committed:
Change start_day and stop_day to start and stop

jeanconn committed:
Vectorize setting up days to check

jeanconn committed:
Improve some variable names

jeanconn committed:
Remove separate characteristics file

jeanconn commented on aca_required_temp.py:
```diff
@@ -0,0 +1,189 @@
+                                                   day_pitch,
+                                                   time=day,
+                                                   cone_stars=cone_stars)
+        # should we have values or skip entries for None here?
```
Open question about what to do with really bad fields.

taldcroft replied:
The worst possible field would be one that has no stars brighter than 11.0.  I actually don't think that exists, but even if it does then a table with values that essentially say "can't observe this" will be generated.

This reminds me that we probably also want fields `nom_n_acq` and `best_n_acq` that provide the expected mean number of acq stars acquired.

taldcroft replied:
I called these `n_stars_best` and `n_stars_nom` in the original email.  We just need to decide if `best` and `nom` are in front or at the end.

jeanconn commented on aca_required_temp.py:
```diff
@@ -0,0 +1,189 @@
+N_ACQ_STARS = 5
+EDGE_DIST = 30
+COLD_T_CCD = -21
+WARM_T_CCD = -5
```
Open question about reasonable value for this to start: https://github.com/sot/aca_required_temp/commit/16bcfc3542c0ec7788d335879606d0e39f1e41c9

taldcroft replied:
> Is something like WARM_T_CCD = -5 reasonable for our search? Or should this go warmer to further optimize "best" roll/temp candidate?

Any star field that is OK at -5 C is going to be great and we don't need to further optimize.

taldcroft commented on aca_required_temp.py:
```diff
@@ -0,0 +1,189 @@
+    return ok_cone_stars
+
+
+def select_fov_stars(ra, dec, roll, cone_stars):
```
This is a fine start for just getting things rolling, but eventually we will need more accurate star selection.  From the email spec: "Identify all stars that meet the most liberal star selection requirements in SAUSAGE and pick the 8 brightest".  The point being that a field with 3 very close bright stars can suddenly go from looking great to looking horrible if you apply spoiler rules.  

I don't see a way around coding up a mini-SAUSAGE here since we are not providing statistical guidance of star catalog quality, but rather specific information about each target.

That's a long way of saying that eventually this function should be called `select stars()`.

jeanconn replied:
> That's a long way of saying that  eventually this function should be
> called select stars().

I figured, though I still wasn't completely clear on what you'd like for a
balance of fidelity vs complexity.  I figure we can approach it in steps.

Also wondered about using stats a little bit in asking questions like "how
bad would this catalog be if you just discarded the brightest 3
candidates?".

taldcroft replied:
> Also wondered about using stats a little bit in asking questions like "how
> bad would this catalog be if you just discarded the brightest 3
> candidates?".

? I don't understand why you want to ask that question.  We won't ever arbitrarily discard bright stars, so there needs to be a quantified driver (i.e. spoiled or whatnot).

jeanconn replied:
> ? I don't understand why you want to ask that question. We won't ever arbitrarily discard bright stars, so there needs to be a quantified driver (i.e. spoiled or whatnot).

Not trying to add this question to _this_ code.   Just wondering about
drawing more attention for ourselves to attitudes with less "margin"(?) or
depth of good candidates.

taldcroft replied:
> Not trying to add this question to _this_ code.   Just wondering about drawing more attention for ourselves to attitudes with less "margin"(?) or depth of good candidates.

We'll get that from `n_stars_best`.  We may also want `prob_n_stars_lte_3_best,nom`.  Don't do this for now but just leave open for thought.  I'm working on some analysis to justify the best single :+1: or :-1: metric for a catalog.

taldcroft commented on aca_required_temp.py:
```diff
@@ -0,0 +1,189 @@
+    # take the 8 brightest
+    stars = fov_stars[0:8]
+    id_hash = tuple(stars['AGASC_ID'])
+    if id_hash in TEMP_CACHE:
```
Here you can use the same idiom that is normal for a `property` with a `_var` underneath:

```
if id_hash not in T_CCD_CACHE:
    # Get tuple of (t_ccd, n_acq) for this star field and cache
    T_CCD_CACHE[id_hash] = t_ccd_warm_limit(...)
return T_CCD_CACHE[id_hash]
```

taldcroft commented on aca_required_temp.py:
```diff
@@ -0,0 +1,189 @@
+
+
+def get_rolldev(pitch):
+    if pitch < ROLL_TABLE['pitch'][0]:
```
Should raise an exception since this violates the CARD.  It means the upstream code is doing something wrong.

jeanconn replied:
> Should raise an exception since this violates the CARD. It means the
> upstream code is doing something wrong.

I was referencing your 45 degrees in the email spec and the table starts at
46.

taldcroft replied:
Sorry, the original 45 was sloppy.  It's really 46.0, and in reality probably 46.1 or so.

taldcroft commented on aca_required_temp.py:
```diff
@@ -0,0 +1,189 @@
+def get_rolldev(pitch):
+    if pitch < ROLL_TABLE['pitch'][0]:
+        return 0
+    idx = np.searchsorted(ROLL_TABLE['pitch'], pitch)
```
Use `side='right'`, otherwise a value of 46.0 will return `idx = 0` and hence the wrong value from `ROLL_TABLE`.

taldcroft commented on aca_required_temp.py:
```diff
@@ -0,0 +1,189 @@
+
+    # set the agasc lookup time to be in the middle of the cycle for
+    # proper motion correction
+    agasc_mid_time = DateTime((DateTime(start).secs + DateTime(stop).secs)
```
What is this the "mid time" of?  It's not the AGASC (which doesn't have a formal time, except that by chance almost all have epoch=J2000.  It's the "mid time" of the LTS being planned.  So:

```
# Center time of the long-term schedule (LTS) cycle ...
lts_mid_time = start + (stop - start) / 2
```

(Again using DateTime arithmetic).

taldcroft commented on aca_required_temp.py:
```diff
@@ -0,0 +1,189 @@
+    for day in days.date:
+        nom_roll = nominal_roll(ra, dec, time=day)
+        day_pitch = pitch(ra, dec, time=day)
+        if day_pitch < 45 or day_pitch > 170:
```
Should be `day_pitch < 46.0`.  (The `.0` is obviously not needed but I think it adds clarity in this case by implying that `day_pitch` is a float).

taldcroft commented on aca_required_temp.py:
```diff
@@ -0,0 +1,189 @@
+        if day_pitch < 45 or day_pitch > 170:
+            continue
+
+        nom_roll_t_ccd = max_temp(ra, dec, nom_roll, time=day, cone_stars=cone_stars)
```
Delegate all this logic to a single function `get_t_ccd_roll` instead of having two places where there is a decision about what is best and having two places that call `max_temp`.  This way you kill the `nom_roll = ..` call above and have a single call in this loop:

```
nom_t_ccd, nom_roll, best_t_ccd, best_roll = get_t_ccd_roll(
        ra, dec, pitch, time=day, cone_stars=cone_stars)
rows.append((day, pitch, nom_roll, nom_t_ccd, best_roll, best_t_ccd))
```

Previously you would have done `rows = []` and finish with below (with no need to sort):

```
table = Table(rows=rows, names=(...))
for name in ('pitch', ...):
    table[name].format = '{:.2f}'
```

taldcroft commented on aca_required_temp.py:
```diff
@@ -0,0 +1,189 @@
+    temp_table = temps_for_attitude(opt.ra, opt.dec,
+                                    start=opt.start,
+                                    stop=opt.stop)
+    temp_table.write(opt.out, format='ascii.fixed_width_two_line')
```
We should ask SOT MP what they want.  Fixed width two line is nice to read for humans, and easy to read in Python with io.ascii, but requires special handling for any other platform to machine-read because of the extra line.

jeanconn replied:
Opened as #8

taldcroft commented on README.md:
```diff
@@ -0,0 +1,26 @@
+ACA Required Temperature
```
Should be "maximum" not "required".  The latter either implies the temperature must be exactly the required value, or if that is understand still leaves ambiguity about the directionality.

taldcroft commented on README.md:
```diff
@@ -0,0 +1,26 @@
+------------------------
+
+N Stars == 5
+-15C Mag Limit == 10.0
```
This (-15 C mag limit) is not part of the star field evaluation.

jeanconn replied:
> This is not part of the star field evaluation.

The spec should now be updated to reference a spec from your new code now
that you have implemented the better model and I won't be playing with the
0.15 mag per degree idea.

jeanconn replied:
Keeping this open as #9

taldcroft commented on README.md:
```diff
@@ -0,0 +1,26 @@
+-15C Mag Limit == 10.0
+At least 30 arcsecs from FOV chip edge
+
+Guide Requirements
```
The initial release of the tool for cycle 17 will just be concerned with ACQ stars.  As written, ACQ will always be more stringent than GUIDE so we don't need to reference this.

taldcroft commented on aca_required_temp.py:
```diff
@@ -0,0 +1,189 @@
+
+import agasc
+from Chandra.Time import DateTime
+from Ska.Sun import nominal_roll, pitch
```
Better not to overload common names like `pitch` which is prone to error (or forces using a different name when `pitch` (meaning the pitch value) would be appropriate.  Having `Ska.Sun.pitch(..)` in the code is fine and even a bit more clear.

jeanconn committed:
Leave Ska.Sun imports in that namespace

jeanconn committed:
PEP8 fixes

jeanconn committed:
Remove Guide requirements from readme

jeanconn committed:
Change 'Required' to 'Maximum' temperature

jeanconn committed:
Remove pre-filtering customized 'get_agasc_cone'

Remove customized 'get_agasc_cone' and move star filtering/selection
to a select_stars() routine

jeanconn committed:
Set warm temp checked limit to -10C

jeanconn committed:
Change pitch exclusion low limit from 45 to 46.4

jeanconn committed:
Move t_ccd/roll determination to get_t_ccd_roll routine

jeanconn committed:
Remove handling of out of spec pitch (should error)

jeanconn committed:
Use 'right' side np.searchsorted

The documentation on the 'sides' here is still confusing for me, but per
Tom and testing it looks like this puts it on the 'side' we want in cases
of exactly matching the pitch in the ROLL_TABLE.

jeanconn committed:
Rename temperature cache and be more idiomatic when accessing

jeanconn committed:
Relabel proper motion mid time

Relabel proper motion mid time and use float DateTime arithmetic

jeanconn committed:
Rename main temperature method

jeanconn commented:
May I merge this PR?  I think I've got the open items as new issues, and I don't think there is more to be gained in discussion on this page.

taldcroft commented:
OK, feel free to merge.
